pallets/foreign-investments/src/hooks.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::{
+	investments::{Investment, InvestmentCollector},
+	StatusNotificationHook,
+};
+use cfg_types::investments::{CollectedAmount, ForeignInvestmentInfo};
+use frame_support::transactional;
+use sp_runtime::{DispatchError, DispatchResult};
+use sp_std::marker::PhantomData;
+
+use crate::{
+	errors::{InvestError, RedeemError},
+	types::{InvestState, InvestTransition, RedeemState, RedeemTransition, TokenSwapReason},
+	Config, Error, ForeignInvestmentInfo as ForeignInvestmentInfoStorage, InvestmentState, Pallet,
+	RedemptionState, SwapOf,
+};
+
+/// The hook struct which acts upon a fulfilled swap order. Depending on the
+/// last swap reason, advances either the [`InvestmentState`] or
+/// [`RedemptionState`].
+///
+/// Assumes `TokenSwaps` as caller of of the the `notify_status_change` message.
+pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+
+// Hook execution for (partially) fulfilled token swaps which should be consumed
+// by `TokenSwaps`.
+impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+	type Error = DispatchError;
+	type Id = T::TokenSwapOrderId;
+	type Status = SwapOf<T>;
+
+	#[transactional]
+	fn notify_status_change(
+		id: T::TokenSwapOrderId,
+		status: SwapOf<T>,
+	) -> Result<(), DispatchError> {
+		let maybe_info = ForeignInvestmentInfoStorage::<T>::get(id);
+
+		if maybe_info.is_none() {
+			return Ok(());
+		}
+		let info = maybe_info.expect("Cannot be None");
+
+		let reason = info
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		match reason {
+			TokenSwapReason::Investment => {
+				// If the investment requires to be collected, the transition of the
+				// `InvestState` would fail. By implicitly collecting here, we defend against
+				// that and ensure that the swap order fulfillment won't be reverted (since this
+				// function is `transactional`).
+
+				// NOTE: We only collect the tranche tokens, but do not transfer them back. This
+				// updates the unprocessed investment amount such that transitioning the
+				// `InvestState` is not blocked. The user still has to do that manually by
+				// sending `CollectInvest`.
+				if T::Investment::investment_requires_collect(&info.owner, info.id) {
+					T::Investment::collect_investment(info.owner.clone(), info.id)?;
+				}
```

Comment:
Good point, we could. The minor difference is the weight when exiting early: If `collect_investment` exits early, 2-3 reads were performed whereas `investment_requires_collect` does 1-2.

Reply:
I think it mostly depends on how we will benchmark this, and how complex we will want the benchmark to be.

From the testing perspective, it is easier if we call directly the `collect_investment`, because we have fewer branches to test and few methods to mock. But nothing serious here.